samples/s03-singleAABB/hostCode.cpp

```diff
@@ -0,0 +1,267 @@
+  GPRTContext context = gprtContextCreate();
+  GPRTModule module = gprtModuleCreate(context,s03_deviceCode);
+
+  // ##################################################################
+  // set up all the GPU kernels we want to run
+  // ##################################################################
+
+  // -------------------------------------------------------
+  // declare geometry type
+  // -------------------------------------------------------
+  GPRTVarDecl aabbGeomVars[] = {
+    { /* sentinel to mark end of list */ }
+  };
+  GPRTGeomType aabbGeomType
+    = gprtGeomTypeCreate(context,
+                        GPRT_AABBS, // <- This is new!
+                        sizeof(AABBGeomData),
+                        aabbGeomVars,-1);
+  gprtGeomTypeSetClosestHitProg(aabbGeomType,0,
+                           module,"AABBClosestHit");
+  gprtGeomTypeSetIntersectionProg(aabbGeomType,0,
+                           module,"AABBIntersection");
+
+  // -------------------------------------------------------
+  // set up miss
+  // -------------------------------------------------------
+  GPRTVarDecl missVars[]
+    = {
+    { "color0", GPRT_FLOAT3, GPRT_OFFSETOF(MissProgData,color0)},
+    { "color1", GPRT_FLOAT3, GPRT_OFFSETOF(MissProgData,color1)},
+    { /* sentinel to mark end of list */ }
+  };
+  GPRTMiss miss
+    = gprtMissCreate(context,module,"miss",sizeof(MissProgData),
+                        missVars,-1);
+
+  // -------------------------------------------------------
+  // set up ray gen program
+  // -------------------------------------------------------
+  GPRTVarDecl rayGenVars[] = {
+    { "fbSize",        GPRT_INT2,   GPRT_OFFSETOF(RayGenData,fbSize)},
+    { "fbPtr",         GPRT_BUFFER, GPRT_OFFSETOF(RayGenData,fbPtr)},
+    { "world",         GPRT_ACCEL,  GPRT_OFFSETOF(RayGenData,world)},
+    { "camera.pos",    GPRT_FLOAT3, GPRT_OFFSETOF(RayGenData,camera.pos)},
+    { "camera.dir_00", GPRT_FLOAT3, GPRT_OFFSETOF(RayGenData,camera.dir_00)},
+    { "camera.dir_du", GPRT_FLOAT3, GPRT_OFFSETOF(RayGenData,camera.dir_du)},
+    { "camera.dir_dv", GPRT_FLOAT3, GPRT_OFFSETOF(RayGenData,camera.dir_dv)},
+    { /* sentinel to mark end of list */ }
+  };
+  GPRTRayGen rayGen
+    = gprtRayGenCreate(context,module,"simpleRayGen",
+                      sizeof(RayGenData),
+                      rayGenVars,-1);
+  // ##################################################################
+  // set the parameters for those kernels
+  // ##################################################################
+
+  // Setup pixel frame buffer
+  GPRTBuffer frameBuffer
+    = gprtDeviceBufferCreate(context,GPRT_INT,fbSize.x*fbSize.y);
+  gprtRayGenSetBuffer(rayGen,"fbPtr", frameBuffer);
+  gprtRayGenSet2iv(rayGen,"fbSize", (int32_t*)&fbSize);
+
+  // Miss program checkerboard background colors
+  gprtMissSet3f(miss,"color0",0.1f,0.1f,0.1f);
+  gprtMissSet3f(miss,"color1",.0f,.0f,.0f);
```

Comment:
Can we move these closer to the data structure's definition?

Reply:
I moved this lower since other structure parameters can’t be set right away (eg, trees must be built first). Made more sense to me to break things up into declarations first, then a phase of setting parameters.

Reply:
I suppose i could restructure things though so that we could have the parameter sets next to the declaration

Reply:
IMO keeping those together would help learners relate the data structure to its use.

Reply:
In retrospect, this feels like a bad idea. It is often not possible to set parameters right next to the declaration, and makes explaining certain concepts like buffers and the shader binding table harder.

Reply:
![image](https://user-images.githubusercontent.com/13104331/207425237-203f8049-9f68-4a3f-94e7-7dcac86d2a76.png)

For example, here, we can't easily set all ray generation parameters, because this depends on all buffers being allocated, all acceleration structures referenced being constructed, etc, which I'd prefer not talking about until later on in the example. 

We also cannot call gprtBuildShaderBindingTable until gprtBuildPipeline is called, because gprtBuildShaderBindingTable will insert shader records generated by gprtBuildPipeline into the table. gprtBuildPipeline cannot be called until all programs have been declared though, and so we'd need to decouple setting the raygen set parameters calls from the call to upload these parameters to the device.

Reply:
Talking one-on-one, the real issue here is that we're using strings to set these parameters, and the strings are a bit confusing. Ideally we would be using some compile-time mechanism to set these parameters, rather than using a dictionary lookup at runtime.